diff -Naur linux-cff7c42.orig/arch/arm/boot/dts/amlogic/meson8_k200b_1G_emmc_sdhc.dtd linux-cff7c42/arch/arm/boot/dts/amlogic/meson8_k200b_1G_emmc_sdhc.dtd
--- linux-cff7c42.orig/arch/arm/boot/dts/amlogic/meson8_k200b_1G_emmc_sdhc.dtd	2016-08-15 12:09:06.019109805 +0200
+++ linux-cff7c42/arch/arm/boot/dts/amlogic/meson8_k200b_1G_emmc_sdhc.dtd	2016-08-15 12:09:56.447106988 +0200
@@ -142,11 +142,11 @@
 		compatible = "amlogic,mesonfb";
 		dev_name = "mesonfb";
 		status = "okay";
-		reserve-memory = <0x01800000  0x00100000>;
+		reserve-memory = <0x04800000 0x00100000>;
 		reserve-iomap = "true";
                 vmode = <3>; /**0:VMODE_720P 1:VMODE_LCD  2:VMODE_LVDS_1080P 3:VMODE_1080P*/
 		scale_mode = <1>; /*0:default 1:new*/
-		4k2k_fb = <0>;
+		4k2k_fb = <1>;
 		display_size_default = <1920 1080 1920 3240 32>; //1920*1080*4*3 = 0x17BB000
 	};
 
@@ -184,7 +184,7 @@
                 compatible = "amlogic,vdec";
                 dev_name = "vdec.0";
                 status = "okay";
-								reserve-memory = <0x04000000>; // 64M
+								reserve-memory = <0x10000000>;
 //		reserve-iomap = "true";
 		linux,contiguous-region = <&mem_4k2k_decoder>;
         };
diff -Naur linux-cff7c42.orig/arch/arm/boot/dts/amlogic/meson8_k200b_1G_emmc_sdio.dtd linux-cff7c42/arch/arm/boot/dts/amlogic/meson8_k200b_1G_emmc_sdio.dtd
--- linux-cff7c42.orig/arch/arm/boot/dts/amlogic/meson8_k200b_1G_emmc_sdio.dtd	2016-08-15 12:09:06.019109805 +0200
+++ linux-cff7c42/arch/arm/boot/dts/amlogic/meson8_k200b_1G_emmc_sdio.dtd	2016-08-15 12:09:56.447106988 +0200
@@ -142,11 +142,11 @@
 		compatible = "amlogic,mesonfb";
 		dev_name = "mesonfb";
 		status = "okay";
-		reserve-memory = <0x01800000  0x00100000>;
+		reserve-memory = <0x04800000 0x00100000>;
 		reserve-iomap = "true";
                 vmode = <3>; /**0:VMODE_720P 1:VMODE_LCD  2:VMODE_LVDS_1080P 3:VMODE_1080P*/
 		scale_mode = <1>; /*0:default 1:new*/
-		4k2k_fb = <0>;
+		4k2k_fb = <1>;
 		display_size_default = <1920 1080 1920 3240 32>; //1920*1080*4*3 = 0x17BB000
 	};
 
@@ -184,7 +184,7 @@
                 compatible = "amlogic,vdec";
                 dev_name = "vdec.0";
                 status = "okay";
-								reserve-memory = <0x04000000>; // 64M
+								reserve-memory = <0x10000000>;
 //		reserve-iomap = "true";
 		linux,contiguous-region = <&mem_4k2k_decoder>;
         };
@@ -1063,6 +1063,24 @@
             511      60  70  75  80
         >;
 	};
+/// ***************************************************************************************
+///    -       Led
+//$$ MODULE="Led"
+//$$ DEVICE="led"
+//$$ L2 PROP_STR = "status"
+//$$ L2 PROP_U32 = "board_type"
+//$$ L2 PROP_STR = "red_led_gpio"
+//$$ L2 PROP_STR = "blue_led_gpio"
+//$$ L2 PROP_U32 = "led_mode"
+	led{
+		compatible = "amlogic-led";
+		dev_name = "aml-setio";
+		status = "okay";
+		red_led_gpio = "GPIO_TEST_N";/** if your board just need one gpio,please config red_led_gpio and blue_led_gpio the same */
+		blue_led_gpio = "GPIO_TEST_N";/**  */
+		red_led_mode = <1>; /** 1:H on L off , 0:H off L on */
+		blue_led_mode= <1>; /** 1:H on L off , 0:H off L on */
+	};
 
 /// ***************************************************************************************
 /// -   Audio
diff -Naur linux-cff7c42.orig/arch/arm/boot/dts/amlogic/meson8_k200b_2G_emmc_sdhc.dtd linux-cff7c42/arch/arm/boot/dts/amlogic/meson8_k200b_2G_emmc_sdhc.dtd
--- linux-cff7c42.orig/arch/arm/boot/dts/amlogic/meson8_k200b_2G_emmc_sdhc.dtd	2016-08-15 12:09:06.019109805 +0200
+++ linux-cff7c42/arch/arm/boot/dts/amlogic/meson8_k200b_2G_emmc_sdhc.dtd	2016-08-15 12:09:56.447106988 +0200
@@ -133,11 +133,11 @@
 		compatible = "amlogic,mesonfb";
 		dev_name = "mesonfb";
 		status = "okay";
-		reserve-memory = <0x01800000  0x00100000>;
+		reserve-memory = <0x04800000 0x00100000>;
 		reserve-iomap = "true";
                 vmode = <3>; /**0:VMODE_720P 1:VMODE_LCD  2:VMODE_LVDS_1080P 3:VMODE_1080P*/
 		scale_mode = <1>; /*0:default 1:new*/
-		4k2k_fb = <0>;
+		4k2k_fb = <1>;
 		display_size_default = <1920 1080 1920 3240 32>; //1920*1080*4*3 = 0x17BB000
 	};
 
@@ -175,7 +175,7 @@
                 compatible = "amlogic,vdec";
                 dev_name = "vdec.0";
                 status = "okay";
-								reserve-memory = <0x04000000>; // 64M
+								reserve-memory = <0x10000000>;
 //		reserve-iomap = "true";
         };
 ///	-	picture decoder
@@ -1054,6 +1054,24 @@
             511      60  70  75  80
         >;
 	};
+/// ***************************************************************************************
+///    -       Led
+//$$ MODULE="Led"
+//$$ DEVICE="led"
+//$$ L2 PROP_STR = "status"
+//$$ L2 PROP_U32 = "board_type"
+//$$ L2 PROP_STR = "red_led_gpio"
+//$$ L2 PROP_STR = "blue_led_gpio"
+//$$ L2 PROP_U32 = "led_mode"
+	led{
+		compatible = "amlogic-led";
+		dev_name = "aml-setio";
+		status = "okay";
+		red_led_gpio = "GPIO_TEST_N";/** if your board just need one gpio,please config red_led_gpio and blue_led_gpio the same */
+		blue_led_gpio = "GPIO_TEST_N";/**  */
+		red_led_mode = <1>; /** 1:H on L off , 0:H off L on */
+		blue_led_mode= <1>; /** 1:H on L off , 0:H off L on */
+	};
 
 /// ***************************************************************************************
 /// -   Audio
diff -Naur linux-cff7c42.orig/arch/arm/boot/dts/amlogic/meson8m2_n200_1G.dtd linux-cff7c42/arch/arm/boot/dts/amlogic/meson8m2_n200_1G.dtd
--- linux-cff7c42.orig/arch/arm/boot/dts/amlogic/meson8m2_n200_1G.dtd	2016-08-15 12:09:06.023109805 +0200
+++ linux-cff7c42/arch/arm/boot/dts/amlogic/meson8m2_n200_1G.dtd	2016-08-15 12:21:22.000000000 +0200
@@ -140,11 +140,11 @@
 		compatible = "amlogic,mesonfb";
 		dev_name = "mesonfb";
 		status = "okay";
-		reserve-memory = <0x01800000  0x00100000>;
+		reserve-memory = <0x04800000 0x00100000>;
 		reserve-iomap = "true";
                 vmode = <3>; /**0:VMODE_720P 1:VMODE_LCD  2:VMODE_LVDS_1080P 3:VMODE_1080P*/
 		scale_mode = <1>; /*0:default 1:new*/
-		4k2k_fb = <0>;
+		4k2k_fb = <1>;
 		display_size_default = <1920 1080 1920 3240 32>; //1920*1080*4*3 = 0x17BB000
 	};
 
@@ -182,7 +182,7 @@
                 compatible = "amlogic,vdec";
                 dev_name = "vdec.0";
                 status = "okay";
-								reserve-memory = <0x04000000>; // 64M
+								reserve-memory = <0x10000000>;
 //		reserve-iomap = "true";
 		linux,contiguous-region = <&mem_4k2k_decoder>;
         };
@@ -1932,15 +1932,21 @@
 	};
 /// ***************************************************************************************
 ///	-	Ethernet
-        ethmac: ethernet@0xc9410000{
-                compatible = "amlogic,meson8m2-rmii-dwmac", "snps,dwmac";
-                reg = <0xc9410000 0x10000
-                       0xc1108108 0x4>;
-                interrupts = <0 8 1>;
-                interrupt-names = "macirq";
-                clocks = "clk81";
-                clock-names = "stmmaceth";
-        };
+//$$ MODULE="Ethernet"
+//$$ DEVICE="meson-eth"
+//$$ L2 PROP_STR = "status"
+    meson-eth{
+    compatible = "amlogic,meson-eth";
+    dev_name = "meson-eth";
+    status = "okay";
+    ethbaseaddr = <0xfe0c0000>;
+    interruptnum = <40>;
+    new_maclogic = <1>;
+    phy_interface = <1>;
+    reset_pin_enable = <0>;
+    reset_pin = "GPI0H_4";
+    reset_delay = <100>;
+	};
 
 /// ***************************************************************************************
 ///	-	DVB
diff -Naur linux-cff7c42.orig/arch/arm/boot/dts/amlogic/meson8m2_n200_2G.dtd linux-cff7c42/arch/arm/boot/dts/amlogic/meson8m2_n200_2G.dtd
--- linux-cff7c42.orig/arch/arm/boot/dts/amlogic/meson8m2_n200_2G.dtd	2016-08-15 12:09:06.023109805 +0200
+++ linux-cff7c42/arch/arm/boot/dts/amlogic/meson8m2_n200_2G.dtd	2016-08-15 12:21:48.000000000 +0200
@@ -133,11 +133,11 @@
 		compatible = "amlogic,mesonfb";
 		dev_name = "mesonfb";
 		status = "okay";
-		reserve-memory = <0x01800000  0x00100000>;
+		reserve-memory = <0x04800000 0x00100000>;
 		reserve-iomap = "true";
                 vmode = <3>; /**0:VMODE_720P 1:VMODE_LCD  2:VMODE_LVDS_1080P 3:VMODE_1080P*/
 		scale_mode = <1>; /*0:default 1:new*/
-		4k2k_fb = <0>;
+		4k2k_fb = <1>;
 		display_size_default = <1920 1080 1920 3240 32>; //1920*1080*4*3 = 0x17BB000
 	};
 
@@ -175,7 +175,7 @@
                 compatible = "amlogic,vdec";
                 dev_name = "vdec.0";
                 status = "okay";
-								reserve-memory = <0x04000000>; // 64M
+								reserve-memory = <0x10000000>;
 //		reserve-iomap = "true";
         };
 ///	-	picture decoder
@@ -1060,6 +1060,24 @@
             511      60  70  75  80
         >;
 	};
+/// ***************************************************************************************
+///    -       Led
+//$$ MODULE="Led"
+//$$ DEVICE="led"
+//$$ L2 PROP_STR = "status"
+//$$ L2 PROP_U32 = "board_type"
+//$$ L2 PROP_STR = "red_led_gpio"
+//$$ L2 PROP_STR = "blue_led_gpio"
+//$$ L2 PROP_U32 = "led_mode"
+	led{
+		compatible = "amlogic-led";
+		dev_name = "aml-setio";
+		status = "okay";
+		red_led_gpio = "GPIO_TEST_N";/** if your board just need one gpio,please config red_led_gpio and blue_led_gpio the same */
+		blue_led_gpio = "GPIO_TEST_N";/**  */
+		red_led_mode = <1>; /** 1:H on L off , 0:H off L on */
+		blue_led_mode= <1>; /** 1:H on L off , 0:H off L on */
+	};
 
 /// ***************************************************************************************
 /// -   Audio
@@ -1905,15 +1923,21 @@
 	};
 /// ***************************************************************************************
 ///	-	Ethernet
-        ethmac: ethernet@0xc9410000{
-                compatible = "amlogic,meson8m2-rmii-dwmac", "snps,dwmac";
-                reg = <0xc9410000 0x10000
-                       0xc1108108 0x4>;
-                interrupts = <0 8 1>;
-                interrupt-names = "macirq";
-                clocks = "clk81";
-                clock-names = "stmmaceth";
-        };
+//$$ MODULE="Ethernet"
+//$$ DEVICE="meson-eth"
+//$$ L2 PROP_STR = "status"
+    meson-eth{
+    compatible = "amlogic,meson-eth";
+    dev_name = "meson-eth";
+    status = "okay";
+    ethbaseaddr = <0xfe0c0000>;
+    interruptnum = <40>;
+    new_maclogic = <1>;
+    phy_interface = <1>;
+    reset_pin_enable = <0>;
+    reset_pin = "GPI0H_4";
+    reset_delay = <100>;
+	};
 
 /// ***************************************************************************************
 ///	-	DVB
